src/doorman/IdAllocator.hpp

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+class IdAllocator {
+    class ReservationObj {
+        IdAllocator &allocator_;
+        uint32_t id_;
+        friend IdAllocator;
+        ReservationObj(IdAllocator &allocator, uint32_t id) : allocator_(allocator), id_(id) {}
+
+    public:
+        ~ReservationObj() { allocator_.release(id_); }
+        ReservationObj(const ReservationObj &) = delete;
+        ReservationObj(ReservationObj &&) = delete;
+        ReservationObj &operator=(const ReservationObj &) = delete;
+        ReservationObj &operator=(ReservationObj &&) = delete;
+
+        [[nodiscard]] uint32_t id() const noexcept { return id_; }
+    };
+
+public:
+    using Reservation = std::shared_ptr<ReservationObj>;
```

Comment:
This is an interesting convention:  rather than having reserve() state that it returns shared_ptr, it's aliased a "Reservation".  It makes the code shorter so I guess it's more readable. Is that a widespread convention these days?

Reply:
It's fairly common. Callers don't need to care that this is a `shared_ptr`, in fact I was going to wrap it in a `Reservation` object that held a shared_ptr, but I was too lazy. Down the line one might wish to do that.

In general, for what it's worth, shared_ptr is considered an anti-pattern. In this case I'm using it very specifically to get the reference counting. Most C++ code gets by without it. It's a common thing you see in code written by folks new to C++ coming from managed languages, though.